src/platform/iotjs_module_adc-linux-general.inl.h

```diff
@@ -16,28 +16,95 @@
 #ifndef IOTJS_MODULE_ADC_LINUX_GENERAL_INL_H
 #define IOTJS_MODULE_ADC_LINUX_GENERAL_INL_H
 
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
 
+#include "iotjs_systemio-linux.h"
 #include "module/iotjs_module_adc.h"
 
 
+#define ADC_INTERFACE "/sys/devices/12d10000.adc/iio:device0/"
+#define ADC_PIN_INTERFACE "in_voltage%d_raw"
+#define ADC_PIN_FORMAT ADC_INTERFACE ADC_PIN_INTERFACE
+
+#define ADC_PATH_BUFFER_SIZE DEVICE_IO_PATH_BUFFER_SIZE
+#define ADC_PIN_BUFFER_SIZE DEVICE_IO_PIN_BUFFER_SIZE
+#define ADC_VALUE_BUFFER_SIZE 64
+
+
+// Implementation used here are based on:
+//  https://www.kernel.org/doc/Documentation/adc/sysfs.txt
+
+
+void InitializeadcWorker(uv_work_t* work_req) {
+  iotjs_adc_reqwrap_t* req_wrap = iotjs_adc_reqwrap_from_request(work_req);
+  iotjs_adc_reqdata_t* req_data = iotjs_adc_reqwrap_data(req_wrap);
+  DDDLOG("ADC InitializeadcWorker()");
+
+  // Check if ADC interface exits.
+  if (DeviceCheckPath(ADC_INTERFACE)) {
+    req_data->result = kAdcErrOk;
+  } else {
+    req_data->result = kAdcErrExport;
+  }
+}
+
+void ReadadcWorker(uv_work_t* work_req) {
+  iotjs_adc_reqwrap_t* req_wrap = iotjs_adc_reqwrap_from_request(work_req);
+  iotjs_adc_reqdata_t* req_data = iotjs_adc_reqwrap_data(req_wrap);
+  uint32_t pin = req_data->pin;
+  DDDLOG("adc ReadadcWorker() - pin: %d", pin);
+
+  char value_path[ADC_PATH_BUFFER_SIZE] = { 0 };
+  snprintf(value_path, sizeof(value_path), ADC_PIN_FORMAT, pin);
+
+  char buffer[ADC_VALUE_BUFFER_SIZE];
+  if (DeviceOpenReadClose(value_path, buffer, sizeof(buffer))) {
+    req_data->result = kAdcErrOk;
+    req_data->value = atoi(buffer);
+  } else {
+    req_data->result = kAdcErrRead;
+  }
+}
+
+
+
 void iotjs_adc_export_worker(uv_work_t* work_req) {
-  IOTJS_ASSERT(!"Not implemented");
+  InitializeadcWorker(work_req);
```

Comment:
Thank you for all these helpful comments.

I would like to keep this ADC implementation interface to be identical like for Nuttx but still keep it similar to GPIO for Linux for further refactoring.

It is clear that ADC, GPIO and PWM for Linux should be implemented 100% in JavaScript using FS API. I love to do it just now, but first I have to test all key APIs at Artik board. It is better to create working API, make tests and next refactor it to make it shine.